packages/core/test/generator.test.ts

```diff
@@ -0,0 +1,38 @@
+import * as expect from 'expect';
+import { describe, it } from 'mocha';
+import { run, sleep, Operation } from '../src/index';
+
+describe('generators', () => {
+
+  function *add(left: number, right: number): Operation<Record<never, never>> {
+    yield sleep(10);
+    let sum =
+      (yield Promise.resolve(left)) + (yield Promise.resolve(right));
+
+    return { sum };
+  }
+
+  it('can run as operations', async () => {
+    expect(await run(add(2, 2))).toEqual({ sum: 4 });
+  });
+
+  it('yields the same thing when run twice in the same operation', async() => {
+    let result = add(2, 2);
+    await run(function*() {
+      expect(yield result).toBe(yield result);
+    })
```

Comment:
I still see this as somewhat of an anti-pattern and would kind of prefer this to throw an error, instead of using the cached promise. I feel like if `add` would have any side-effects, especially if those side effects are dependent on the context that it runs in, then this behaviour becomes a bit of a foot-gun.

Reply:
Yeah, that makes sense. I was also thinking that we can make a linting rule that would complain if you were doing anything but yielding to an iterator or a promise.

Reply:
@jnicklas Updated with a "fail fast" strategy